app/views/PaymentHistory.scala.html

```diff
@@ -42,12 +42,21 @@
 btaNavPartial: Option[Html] = None, origin: Option[String] = None, MFACreditsEnabled: Boolean = true)(implicit request: Request[_], messages: Messages)
 @import implicitDateFormatter.longDate
 
+@sortPayments(payments: List[Payment]) = @{
+            payments
+                .map(payment => if (payment.validMFACreditDescription() ) ( Some(payment.documentDate), payment) else (payment.dueDate, payment) )
+                .map(rec => ( LocalDate.parse(rec._1.getOrElse(throw MissingFieldException("Payment/Credits::Sorting"))).toEpochDay, rec._2) )
+                .sortWith( (left,right) => left._1 < right._1)
+                .map(_._2)
+}
+
 @groupedPayments = @{
     payments.groupBy[Int]{payment => {
             if (payment.validMFACreditDescription()) LocalDate.parse(payment.documentDate).getYear
             else LocalDate.parse(payment.dueDate.getOrElse(throw MissingFieldException("Payment Date"))).getYear
         }
     }.toList.sortBy(_._1).reverse
+      .map(rec => (rec._1, sortPayments(rec._2)))
```

Comment:
Might be nice to use case match here to give the tuple params some context:
.map { case(year, payments) => (year, sortPayments(payments)) }